File: python/days/day5.py
def day5():
    with open('../input/input5.txt', 'r') as f:
        data = f.read().split('\n')[:-1]

    c = [list(reversed([d[4 * i + 1] if d[4 * i + 1] != ' ' else None for d in data[:8]])) for i in range(9)]
    crates9k = [[x for x in row if x is not None] for row in c]
    crates9k1 = [[x for x in row if x is not None] for row in c]
    moves = [(int(row.split(' ')[1]), int(row.split(' ')[3]) - 1, int(row.split(' ')[5]) - 1) for row in data[10:]]

    for m in range(len(moves)):
        qty, src, dst = moves[m]
        crates9k[dst].extend([crates9k[src].pop() for _ in range(qty)])
        crates9k1[dst].extend(list(reversed([crates9k1[src].pop() for _ in range(qty)])))

    print('Part 1: ', ''.join([c[-1] for c in crates9k]))
    print('Part 2: ', ''.join([c[-1] for c in crates9k1]))

File: rust/src/aoc2022/day5.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::fs::File;

pub fn day5(path: &String) {

    let data:Vec<String> = File::open(path).read_lines::<String>(1).collect();

    let mut crates9k = (0..9)
        .map(|i| {
            data[0..8]
                .iter()
                .map(|d| d.chars().nth(4 * i + 1).unwrap())
                .filter(|&d| d != ' ')
                .rev()
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();

    let mut crates9k1 = crates9k.clone();

    let moves = data[10..]
        .iter()
        .map(|row| {
            let parts:Vec<&str> = row.split(" ").collect();
            (
                parts[1].parse::<usize>().unwrap(),
                parts[3].parse::<usize>().unwrap() - 1,
                parts[5].parse::<usize>().unwrap() - 1,
            )
        })
        .collect::<Vec<_>>();

    for m in moves {
        let (qty, src, dst) = m;

        let mut stack:Vec<char> = vec![]; 
	    let mut crates9k_dst = crates9k[dst].clone();
	    let mut crates9k1_dst = crates9k1[dst].clone();

	    for _ in 0..qty {
            stack.push(crates9k1[src].pop().unwrap());
	    }
	    for _ in 0..qty {
            crates9k_dst.push(crates9k[src].pop().unwrap());
		    crates9k1_dst.push(stack.pop().unwrap());
	    }
	    
	    crates9k[dst] = crates9k_dst;
	    crates9k1[dst] = crates9k1_dst;
    }

    println!("Part 1: {}", crates9k.iter().map(|c| c.last().unwrap()).collect::<String>());
    println!("Part 2: {}", crates9k1.iter().map(|c| c.last().unwrap()).collect::<String>());
}
